codetransformer/instructions.py

```diff
@@ -0,0 +1,163 @@
+from abc import ABCMeta, abstractmethod
+from dis import opname, opmap, hasjabs, hasjrel, HAVE_ARGUMENT, stack_effect
+
+
+__all__ = ['Instruction'] + list(opmap)
+
+
+class _immutableattr(object):
+    def __init__(self, op):
+        self._op = op
+
+    def __get__(self, instance, owner):
+        return self._op
+
+
+class InstructionMeta(ABCMeta):
+    _marker = object()  # sentinel
+    _type_cache = {}
+
+    def __init__(self, *args, opcode=None):
+        return super().__init__(*args)
+
+    def __new__(mcls, name, bases, dict_, *, opcode=None):
+        if len(bases) != 1:
+            raise TypeError(
+                '{} does not support multiple inheritence'.format(
+                    mcls.__name__,
+                ),
+            )
+        if bases[0] is mcls._marker:
+            return super().__new__(mcls, name, (object,), dict_)
+
+        if opcode not in opmap.values():
+            raise TypeError('Invalid opcode: {}'.format(opcode))
+
+        dict_['opcode'] = _immutableattr(opcode)
+        dict_['absjmp'] = _immutableattr(opcode in hasjabs)
+        dict_['reljmp'] = _immutableattr(opcode in hasjrel)
+        dict_['opname'] = _immutableattr(opname[opcode])
+
+        have_arg = opcode >= HAVE_ARGUMENT
+        dict_['have_arg'] = _immutableattr(have_arg)
+        mcls._type_cache[opcode] = cls = super().__new__(
```

Comment:
I think I wanted to cache this so that instr.from_opcode would cache; I will make it use the cache.